docs/kgenframeworksapidox.1
\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "kgenframeworksapidox 1"
.TH kgenframeworksapidox 1 "2015-07-11"
.SH "NAME"
kgenframeworksapidox \- KDE API documentation generation tools
.SH "VERSION"
.IX Header "VERSION"
5\.13\.0

.SH "USAGE"
.IX Header "USAGE"

usage: kgenframeworksapidox [\-h] [\-\-depdiagram\-dot\-dir DIR] [\-\-title TITLE]
                            [\-\-man-pages] [\-\-qhp] [\-\-searchengine]
                            [\-\-api\-searchbox] [\-\-qtdoc\-dir QTDOC_DIR]
                            [\-\-qtdoc\-link QTDOC_LINK] [\-\-qtdoc\-flatten\-links]
                            [\-\-doxygen DOXYGEN]
                            [\-\-qhelpgenerator QHELPGENERATOR]
                            [\-\-doxdatadir DOXDATADIR] [\-\-keep\-temp\-dirs]
                            frameworksdir

Generate API documentation for the KDE Frameworks

optional arguments:
  \-h, \-\-help            show this help message and exit

sources:
  frameworksdir         Location of the frameworks modules.
  \-\-depdiagram-dot-dir DIR
                        Generate dependency diagrams, using the .dot files
                        from DIR.

output options:
  \-\-title TITLE         String to use for page titles.
  \-\-man-pages           Generate man page documentation.
  \-\-qhp                 Generate Qt Compressed Help documentation.
  \-\-searchengine        Enable Doxygen's search engine feature.
  \-\-api\-searchbox       Enable the API searchbox (only useful for
                        api.kde.org).

Qt documentation:
  \-\-qtdoc-dir QTDOC_DIR
                        Location of (local) Qt documentation; this is searched
                        for tag files to create links to Qt classes.
  \-\-qtdoc\-link QTDOC_LINK
                        Override Qt documentation location for the links in
                        the html files. May be a path or URL.
  \-\-qtdoc\-flatten-links
                        Whether to assume all Qt documentation html files are
                        immediately under QTDOC_LINK (useful if you set
                        QTDOC_LINK to the online Qt documentation). Ignored if
                        QTDOC_LINK is not set.

paths:
  \-\-doxygen DOXYGEN     (Path to) the doxygen executable.
  \-\-qhelpgenerator QHELPGENERATOR
                        (Path to) the qhelpgenerator executable.

misc:
  \-\-doxdatadir DOXDATADIR
                        Location of the HTML header files and support
                        graphics.
  \-\-keep\-temp\-dirs      Do not delete temporary dirs, useful for debugging.

.SH "AUTHORS"
.IX Header "AUTHORS"
This version of \fBkapidox\fR was written by Aurélien Gâteau agateau@kde.org.
.PP
This man-page was created by Scott Kitterman <scott@kitterman.com>.



src/kgenframeworksapidox.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

# Python 2/3 compatibility (NB: we require at least 2.7)
from __future__ import division, absolute_import, print_function, unicode_literals

import argparse, os, shutil, sys

scriptdir = os.path.dirname(os.path.realpath(__file__))
import kgenapidox

def get_tier(yaml_file):
    """Parse the tier out of a yaml file"""
    with open(yaml_file) as f:
        import yaml
        data = yaml.load(f)
        return data['tier']

def generate_group_menu(tiers):
    """Generate a menu for the frameworks"""
    sections = []
    for t in range(1,5):
        frameworks = []
        for fw in tiers[t]:
            rellink = '../../' + fw['outputdir'] + '/html/index.html'
            frameworks.append({
                'href': rellink,
                'name': fw['fancyname']
                })
        sections.append({
            'title': 'Tier ' + str(t),
            'members': frameworks
            })
    return {'group_title': 'Frameworks', 'sections': sections}

def process_toplevel_html_file(inputfile, outputfile, tiers, title,
        api_searchbox=False):
    """Pass a HTML file through the templating filters"""

    frameworks = {}
    gm_sections = []
    for t in range(1,5):
        gm_frameworks = []
        for fw in tiers[t]:
            rellink = fw['outputdir'] + '/html/index.html'
            gm_frameworks.append({
                'href': rellink,
                'name': fw['fancyname']
                })
            frameworks[fw['framework']] = {'link': rellink, 'name': fw['fancyname']}
        gm_sections.append({
            'title': 'Tier ' + str(t),
            'members': gm_frameworks
            })
    group_menu = {'group_title': 'Frameworks', 'sections': gm_sections}

    mapping = {
            'resources': '.',
            'api_searchbox': api_searchbox,
            # steal the doxygen css from one of the frameworks
            # this means that all the doxygen-provided images etc. will be found
            'doxygencss': tiers[1][0]['outputdir'] + '/html/doxygen.css',
            'title': title,
            'breadcrumbs': {
                'entries': [
                    {
                        'href': 'http://api.kde.org/',
                        'text': 'KDE API Reference'
                    }
                    ]
                },
            'group_menu': group_menu,
            'frameworks': frameworks
        }
    import pystache
    renderer = pystache.Renderer(decode_errors='ignore')
    with open(outputfile, 'w') as outf:
        outf.write(renderer.render_path(inputfile, mapping))

def main():
    parser = argparse.ArgumentParser(description='Generate API documentation ' +
            'for the KDE Frameworks')
    parser.add_argument('frameworksdir',
            help='Location of the frameworks modules.')
    parser.add_argument('--title', default='KDE API Documentation',
            help='String to use for page titles.')
    parser.add_argument('--man-pages', action='store_true',
            help='Generate man page documentation.')
    parser.add_argument('--qhp', action='store_true',
            help='Generate Qt Compressed Help documentation.')
    parser.add_argument('--searchengine', action='store_true',
            help="Enable Doxygen's search engine feature.")
    parser.add_argument('--api-searchbox', action='store_true',
            help="Enable the API searchbox (only useful for api.kde.org).")
    parser.add_argument('--no-apidocs-suffix', action='store_false',
            dest='use_suffix',
            help='Call the apidocs folders "<framework>" instead of ' +
                 '"<framework>-apidocs".')
    parser.add_argument('--doxdatadir',
            help='Location of the HTML header files and support graphics.')
    parser.add_argument('--qtdoc-dir',
            help='Location of (local) Qt documentation; this is searched ' +
                 'for tag files to create links to Qt classes.')
    parser.add_argument('--qtdoc-link',
            help='Override Qt documentation location for the links in the ' +
                 'html files.  May be a path or URL.')
    parser.add_argument('--qtdoc-flatten-links', action='store_true',
            help='Whether to assume all Qt documentation html files ' +
                 'are immediately under QTDOC_LINK (useful if you set ' +
                 'QTDOC_LINK to the online Qt documentation).  Ignored ' +
                 'if QTDOC_LINK is not set.')
    parser.add_argument('--doxygen', default='doxygen',
            help='(Path to) the doxygen executable.')
    parser.add_argument('--qhelpgenerator', default='qhelpgenerator',
            help='(Path to) the qhelpgenerator executable.')
    args = parser.parse_args()

    doxdatadir = kgenapidox.find_doxdatadir_or_exit(args.doxdatadir)
    tagfiles = kgenapidox.search_for_tagfiles(
            suggestion = args.qtdoc_dir,
            doclink = args.qtdoc_link,
            flattenlinks = args.qtdoc_flatten_links,
            searchpaths = ['/usr/share/doc/qt5', '/usr/share/doc/qt'])

    if not os.path.isdir(args.frameworksdir):
        print(args.frameworksdir + " is not a directory")
        exit(2)

    framework_list = os.listdir(args.frameworksdir)
    tiers = {1:[],2:[],3:[],4:[]}
    for framework in framework_list:
        fwdir = os.path.join(args.frameworksdir, framework)
        outputdir = framework
        if args.use_suffix:
            outputdir += '-apidocs'
        readme_file = os.path.join(fwdir, 'README.md')
        yaml_file = os.path.join(fwdir, framework + '.yaml')
        if not os.path.isfile(readme_file) or not os.path.isfile(yaml_file):
            print(framework + " is not a framework (README.md or " +
                 framework + ".yaml missing)")
        else:
            # FIXME: option in yaml file to disable docs
            fancyname = kgenapidox.parse_fancyname(readme_file, default=framework)
            tier = get_tier(yaml_file)
            if tier is None:
                print("Could not find tier for " + framework + "!")
            elif tier < 1 or tier > 4:
                print("Invalid tier " + tier + " for " + framework + "!")
            else:
                tiers[tier].append({
                    'framework': framework,
                    'fancyname': fancyname,
                    'srcdir': fwdir,
                    'outputdir': outputdir
                    })

    for t in range(1,5):
        tiers[t] = sorted(tiers[t], key=lambda f: f['fancyname'].lower())

    kgenapidox.copy_dir_contents(os.path.join(doxdatadir,'htmlresource'),'.')

    group_menu = generate_group_menu(tiers)

    def gen_fw_apidocs(fwinfo, rebuild=False):
        kgenapidox.generate_apidocs(
                modulename = fwinfo['framework'],
                fancyname = fwinfo['fancyname'],
                srcdir = fwinfo['srcdir'],
                outputdir = fwinfo['outputdir'],
                doxdatadir = doxdatadir,
                tagfiles = tagfiles,
                man_pages = args.man_pages,
                qhp = args.qhp,
                searchengine = args.searchengine,
                api_searchbox = args.api_searchbox,
                doxygen = args.doxygen,
                qhelpgenerator = args.qhelpgenerator,
                title = args.title,
                resourcedir = '../..',
                template_mapping = {
                    'breadcrumbs': {
                        'entries': [
                            {
                                'href': 'http://api.kde.org/',
                                'text': 'KDE API Reference'
                            },
                            {
                                'href': '../../index.html',
                                'text': 'Frameworks'
                            }
                            ]
                        },
                    'group_menu': group_menu
                    },
                doxyfile_entries = { 'WARN_IF_UNDOCUMENTED': True }
                )
        if not rebuild:
            tagfile = os.path.abspath(
                        os.path.join(
                            fwinfo['outputdir'],
                            'html',
                            fwinfo['framework']+'.tags'))
            tagfiles.append((tagfile,'../../' + fwinfo['outputdir'] + '/html/'))

    process_toplevel_html_file(os.path.join(doxdatadir, 'frameworks.html'), 'index.html',
            title=args.title, tiers=tiers, api_searchbox=args.api_searchbox)

    for t in range(1,5):
        for fwinfo in tiers[t]:
            gen_fw_apidocs(fwinfo)
        if (t >= 3):
            # Rebuild for interdependencies
            # FIXME: can we be cleverer about deps?
            for fwinfo in tiers[t]:
                shutil.rmtree(fwinfo['outputdir'])
                gen_fw_apidocs(fwinfo, rebuild=True)


if __name__ == "__main__":
    main()

